src/utils.ts

```diff
@@ -0,0 +1,564 @@
+        decoded: typedDecodedScript,
+        token: input.token as string,
+      };
+
+      return typedInput;
+    }),
+    outputs: tx.outputs.map((output): Output => {
+      const typedDecodedScript: DecodedScript = {
+        type: output.decoded.type as string,
+        address: output.decoded.address as string,
+        timelock: output.decoded.timelock ? output.decoded.timelock as number : null,
+        value: output.decoded.value ? output.decoded.value as number : null,
+        tokenData: output.decoded.token_data ? output.decoded.token_data as number : null,
+      };
+
+      const typedOutput: Output = {
+        value: output.value as number,
+        tokenData: output.token_data as number,
+        script: output.script as string,
+        decoded: typedDecodedScript,
+        token: output.token as string,
+        spentBy: output.spent_by ? output.spent_by as string : null,
+      };
+
+      return typedOutput;
+    }),
+    parents: tx.parents,
+    tokens: tx.tokens.map((token: any): Token => {
+      const parsedToken: Token = {
+        uid: token.uid as string,
+        name: token.name as string,
+        symbol: token.symbol as string,
+      };
+
+      return parsedToken;
+    }),
+    raw: tx.raw as string,
+  };
+
+  return parsedTx;
+};
+
+/**
+ * Downloads a block from the full_node using the `block_at_height` API
+ *
+ * @param txId - The block txId
+ * @param noCache - Prevents downloading the block from cache as a reorg may have ocurred
+ */
+export const getBlockByTxId = async (txId: string, noCache: boolean = false) => {
+  return downloadTx(txId, noCache);
+};
+
+/**
+ * Syncs to the latest block
+ *
+ * @generator
+ * @yields {StatusEvent} A status event indicating if a block at a height was successfully sent, \
+ * if an error ocurred or if a reorg ocurred.
+ */
+export async function* syncToLatestBlock(): AsyncGenerator<StatusEvent> {
+  const ourBestBlock: Block = await getWalletServiceBestBlock();
+  const fullNodeBestBlock: Block = await getFullNodeBestBlock();
+
+  // Check if our best block is still in the fullnode's chain
+  const ourBestBlockInFullNode = await getBlockByTxId(ourBestBlock.txId, true);
+
+  if (!ourBestBlockInFullNode.success) {
+    yield {
+      type: 'error',
+      success: false,
+      message: 'Could not validate our best block',
+      error: ourBestBlockInFullNode.message,
+    };
+
+    return;
+  }
+
+  const { meta } = ourBestBlockInFullNode;
+
+  if ((meta.voided_by &&
+       meta.voided_by.length &&
+         meta.voided_by.length > 0)) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our best block was voided, we should reorg.',
+    };
+
+    return;
+  }
+
+  if (ourBestBlock.height > meta.height) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our height is higher than the wallet-service\'s height, we should reorg.',
+    };
+
+    return;
+  }
+
+  console.log('Best block is valid.');
```

Comment:
We could log `Downloading {meta.height - outBestBlock.height} blocks...`.

Reply:
Done in e3ae097005dab38dbe371f3f7865ba3c6dd1e710